src/Microsoft.Azure.SignalR.Common/ClientInvocation/TaskCompletionSourceWithCancellation.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using System.Threading;
+
+namespace Microsoft.Azure.SignalR
+{
+    // Modified from https://github.com/dotnet/aspnetcore/blob/v7.0.0-preview.7.22376.6/src/SignalR/common/Shared/ClientResultsManager.cs#L110
+    // Custom TCS type links `cancellationToken.Cancel` to `tcs.SetCanceld()`. Besides, it provides a customized behaviour in `tcs.SetCanceld()`.
+    // This modified version decoupled this class with `ClientResultsManager`, `connectionId` and `invocationId` which are used in `SetCanceled()`
+    // This version uses a general `Action setCanceldAction` instead.
+
+    // Custom TCS type to avoid the extra allocation that would be introduced if we managed the cancellation separately
+    // Also makes it easier to keep track of the CancellationTokenRegistration for disposal
+    internal sealed class TaskCompletionSourceWithCancellation<T> : TaskCompletionSource<T>
```

Comment:
- A customized TCS is still needed
  - Compared with `TaskCompletionSource`, `TaskCompletionSourceWithCancellation`  provides 2 additional features
    - 1 Links `cancellationToken.Cancel` to `tcs.SetCanceld()`
    - 2 Customized behaviour in `tcs.SetCanceld()`
  - If use `canellationToken.Register(()=>taskCompletionSource.SetCancelled())` instead customized TCS, feature 2 won't work.
  - If follow the implementation of `ClientInvocationManager` in runtime, which uses `CancellationTokenSource.CreateLinkedTokenSource(connectionAborted, timeoutTcs.Token)`, we cannot return  `tcs.Task`

- I modified `TaskCompletionSourceWithCancellation` from https://github.com/dotnet/aspnetcore/blob/v7.0.0-preview.7.22376.6/src/SignalR/common/Shared/ClientResultsManager.cs#L110
- This modified version decoupled the class with `ClientResultsManager`, `connectionId` and `invocationId` which are used in `SetCanceled()`. Instead, it uses a general `Action setCanceledAction` in constructor parameters.

Reply:
`CancellationToken` is not enough. The return value of `AddInvocation` must have a result, which is required by  `InvokeConnectionnAsync` in #1687 ans AspNetCore SignalR